aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStep.java

```diff
@@ -0,0 +1,162 @@
+            }
+            return state;
+        }
+        private State() {}
+        private final Map<String,Integer> iota = new HashMap<String,Integer>();
+        synchronized int allocateNumber(String id) {
+            Integer old = iota.get(id);
+            if (old == null) {
+                old = 0;
+            }
+            int number = old + 1;
+            iota.put(id, number);
+            return number;
+        }
+        /** map from {@link #k} to serial form of {@link StepContext} */
+        final Map<String,String> contexts = new HashMap<String,String>();
+        final Map<String,Object> returnValues = new HashMap<String,Object>();
+        final Map<String,Throwable> errors = new HashMap<String,Throwable>();
+        final Set<String> started = new HashSet<String>();
+    }
+
+    private final String id;
+    private final int number;
+
+    @DataBoundConstructor 
+    public SynchronousNonBlockingStep(String id) {
+        this.id = id;
+        number = State.get().allocateNumber(id);
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    private String k() {
+        return id + "/" + number;
+    }
+
+    public StepContext getContext() {
+        return getContext(State.get(), k());
+    }
+
+    private static StepContext getContext(State s, String k) {
+        return (StepContext) Jenkins.XSTREAM.fromXML(s.contexts.get(k));
+    }
+
+    public static void waitForStart(@Nonnull String k, @CheckForNull Run<?,?> b) throws IOException, InterruptedException {
+        State s = State.get();
+        synchronized (s) {
+            while (!s.started.contains(k)) {
+                if (b != null && !b.isBuilding()) {
+                    throw new AssertionError(JenkinsRule.getLog(b));
+                }
+                s.wait(1000);
+            }
+        }
+    }
+
+    public static class StepExecutionImpl extends AbstractSynchronousNonBlockingStepExecution<Void> {
+
+        @Inject(optional=true) 
+        private SynchronousNonBlockingStep step;
+
+        @Override
+        protected Void run() throws Exception {
+            // Send a test message to the listener
+            getContext().get(TaskListener.class).getLogger().println("Test Sync Message");
+
+            State s = State.get();
+            String k = step.k();
+            if (s.returnValues.containsKey(k)) {
+                System.err.println("Immediately running " + k);
+                getContext().onSuccess(s.returnValues.get(k));
+            } else if (s.errors.containsKey(k)) {
+                System.err.println("Immediately failing " + k);
+                getContext().onFailure(s.errors.get(k));
+            } else {
+                System.err.println("Blocking " + k);
+                s.contexts.put(k, Jenkins.XSTREAM.toXML(getContext()));
+            }
+
+            // Let's wait 2 seconds to give time to the main test thread to reach the wait point
+            Thread.sleep(2000);
```

Comment:
Seems like this should not be necessary. I imagine you are lacking some notification.